packages/react-urql/src/hooks/useFragment.ts

```diff
@@ -0,0 +1,339 @@
+   * Hint: This should be wrapped in a `useMemo` hook, to make sure that your
+   * component doesn’t infinitely update.
+   *
+   * @example
+   * ```ts
+   * const result = useFragment({
+   *   query,
+   *   data,
+   *   context: useMemo(() => ({
+   *     suspense: true,
+   *   }), [])
+   * });
+   * ```
+   */
+  context: Partial<OperationContext>;
+  /** A GraphQL document to mask this fragment against.
+   *
+   * @remarks
+   * This Document should contain atleast one FragmentDefinitionNode or
+   * a FragmentDefinitionNode with the same name as the `name` property.
+   */
+  query: GraphQLRequestParams<Data, AnyVariables>['query'];
+  /** A JSON object which we will extract properties from to get to the
+   * masked fragment.
+   */
+  data: Data;
+  /** An optional name of the fragment to use from the passed Document. */
+  name?: string;
+};
+
+/** State of the current query, your {@link useFragment} hook is executing.
+ *
+ * @remarks
+ * `UseFragmentState` is returned by {@link useFragment} and
+ * gives you the masked data for the fragment.
+ */
+export interface UseFragmentState<Data = any> {
+  /** Indicates whether `useFragment` is waiting for a new result.
+   *
+   * @remarks
+   * When `useFragment` is passed a new query and/or variables, it will
+   * start executing the new query operation and `fetching` is set to
+   * `true` until a result arrives.
+   */
+  fetching: boolean;
+  /** The {@link OperationResult.data} for the masked fragment. */
+  data?: Data;
+}
+
+const isSuspense = (client: Client, context?: Partial<OperationContext>) =>
+  context && context.suspense !== undefined
+    ? !!context.suspense
+    : client.suspense;
+
+/** Hook to mask a GraphQL Fragment given its data. (BETA)
+ *
+ * @param args - a {@link UseFragmentArgs} object, to pass a `fragment` and `data`.
+ * @returns a {@link UseFragmentState} result.
+ *
+ * @remarks
+ * `useFragments` allows GraphQL fragments to mask their data.
+ * Given {@link UseFragmentArgs.query} and {@link UseFragmentArgs.data}, it will
+ * return the masked data for the fragment contained in query.
+ *
+ * Additionally, if the `suspense` option is enabled on the `Client`,
+ * the `useFragment` hook will suspend instead of indicating that it’s
+ * waiting for a result via {@link UseFragmentState.fetching}.
+ *
+ * @example
+ * ```ts
+ * import { gql, useFragment } from 'urql';
+ *
+ * const TodoFields = gql`
+ *   fragment TodoFields on Todo { id name }
+ * `;
+ *
+ * const Todo = (props) => {
+ *   const result = useQuery({
+ *     data: props.todo,
+ *     query: TodoFields,
+ *     variables: {},
+ *   });
+ *   // ...
+ * };
+ * ```
+ */
+export function useFragment<
+  Data extends Record<string, any> = Record<string, any>,
+>(args: UseFragmentArgs<Data>): UseFragmentState<Data> {
+  const client = useClient();
+  const cache = getCacheForClient(client);
+  const suspense = isSuspense(client, args.context);
+  const request = useRequest(args.query, args.data);
```

Comment:
Not sure whether this is too expensive, we could just rely on `args.data` and `args.query`